### PR TITLE
[NCL-8062] Artifact Product Milestones endpoint is unfiltrable and unsortable

### PR DESCRIPTION
**Pull request by akridl**
### Checklist:

* [x] Have you added unit tests for your change?

### CODE REVIEW

File: facade/src/main/java/org/jboss/pnc/facade/providers/ProductMilestoneProviderImpl.java [hunk 1]
```diff
@@ -253,7 +270,11 @@ public Page<MilestoneInfo> getMilestonesOfArtifact(String artifactId, int pageIn
             return new Page<>();
         }
 
-        CriteriaQuery<Tuple> query = milestoneInfoQuery(cb, milestoneIds);
+        Predicate<org.jboss.pnc.model.ProductMilestone> rsqlPredicate = rsqlPredicateProducer
+                .getCriteriaPredicate(milestoneInfoRSQLMapper, queryRsql);
```
Comment (janinko) [inline facade/src/main/java/org/jboss/pnc/facade/providers/ProductMilestoneProviderImpl.java:274]: could the filtering be moved to getMilestoneIdByBuildRecord and getDependentMilestoneIds (or could it be all united to single query? But maybe keep it rather nice then optimized unless it's slow)

File: facade/src/main/java/org/jboss/pnc/facade/rsql/RSQLProducer.java [hunk 2]
```diff
@@ -40,6 +41,16 @@ public interface RSQLProducer {
      */
     <DB extends GenericEntity<?>> Predicate<DB> getCriteriaPredicate(Class<DB> type, String rsql);
 
+    /**
+     * Analogical to {@link this#getCriteriaPredicate(Class, String)}. Unlike the previous method, it uses only this
+     * single mapper during the whole computation.
```
Comment (janinko) [inline facade/src/main/java/org/jboss/pnc/facade/rsql/RSQLProducer.java:46]: what if the mapper provided would reference other mappers? How does it guarantee that no other mappers are used?
fix the javadoc.

File: facade/src/main/java/org/jboss/pnc/facade/rsql/RSQLProducer.java [hunk 3]
```diff
@@ -60,6 +71,16 @@ public interface RSQLProducer {
      */
     <DB extends GenericEntity<?>> SortInfo<DB> getSortInfo(Class<DB> type, String rsql);
 
+    /**
+     * Analogical to {@link this#getSortInfo(Class, String)}. Unlike the previous method, it uses only this single
+     * mapper during the whole computation.
```
Comment (janinko) [inline facade/src/main/java/org/jboss/pnc/facade/rsql/RSQLProducer.java:76]: ditto

**Review by janinko — CHANGES_REQUESTED** [review]

File: facade/src/main/java/org/jboss/pnc/facade/providers/ProductMilestoneProviderImpl.java (hunk 1 above)
Comment (akridl) [reply facade/src/main/java/org/jboss/pnc/facade/providers/ProductMilestoneProviderImpl.java:274]: LazyInitializationException

**Review by janinko — APPROVED** [review]

File: facade/src/main/java/org/jboss/pnc/facade/rsql/RSQLProducerImpl.java [hunk 4]
```diff
@@ -96,12 +94,22 @@ public RSQLProducerImpl() {
 
     @Override
     public <DB extends GenericEntity<?>> Predicate<DB> getCriteriaPredicate(Class<DB> type, String rsql) {
+        // TODO: Uncomment these 3 lines once NCL-8692, NCL-8693
```
Comment (janinko) [inline facade/src/main/java/org/jboss/pnc/facade/rsql/RSQLProducerImpl.java:97]:
```suggestion
        // TODO: Remove these 3 lines once NCL-8692, NCL-8693 are fixed
```
ditto one down

**Review by janinko — CHANGES_REQUESTED** [review]

**Review by janinko — APPROVED** [review]

**Pull request merged**